pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceAssignmentDriver.java

```diff
@@ -55,40 +55,66 @@ public InstanceAssignmentDriver(TableConfig tableConfig) {
+      @Nullable InstancePartitions preConfiguredInstancePartitions, @Nullable Boolean minimizeDataMovement) {
     String tableNameWithType = _tableConfig.getTableName();
     InstanceAssignmentConfig assignmentConfig =
         InstanceAssignmentConfigUtils.getInstanceAssignmentConfig(_tableConfig, instancePartitionsType);
     return getInstancePartitions(
         instancePartitionsType.getInstancePartitionsName(TableNameBuilder.extractRawTableName(tableNameWithType)),
-        assignmentConfig, instanceConfigs, existingInstancePartitions, preConfiguredInstancePartitions);
+        assignmentConfig, instanceConfigs, existingInstancePartitions, preConfiguredInstancePartitions,
+        minimizeDataMovement);
   }
 
   public InstancePartitions assignInstances(String tierName, List<InstanceConfig> instanceConfigs,
       @Nullable InstancePartitions existingInstancePartitions, InstanceAssignmentConfig instanceAssignmentConfig) {
+    return assignInstances(tierName, instanceConfigs, existingInstancePartitions, instanceAssignmentConfig, null);
+  }
+
+  public InstancePartitions assignInstances(String tierName, List<InstanceConfig> instanceConfigs,
+      @Nullable InstancePartitions existingInstancePartitions, InstanceAssignmentConfig instanceAssignmentConfig,
+      @Nullable Boolean minimizeDataMovement) {
     return getInstancePartitions(
         InstancePartitionsUtils.getInstancePartitionsNameForTier(_tableConfig.getTableName(), tierName),
-        instanceAssignmentConfig, instanceConfigs, existingInstancePartitions, null);
+        instanceAssignmentConfig, instanceConfigs, existingInstancePartitions, null, minimizeDataMovement);
   }
 
   private InstancePartitions getInstancePartitions(String instancePartitionsName,
       InstanceAssignmentConfig instanceAssignmentConfig, List<InstanceConfig> instanceConfigs,
       @Nullable InstancePartitions existingInstancePartitions,
-      @Nullable InstancePartitions preConfiguredInstancePartitions) {
+      @Nullable InstancePartitions preConfiguredInstancePartitions, @Nullable Boolean minimizeDataMovementOverride) {
     String tableNameWithType = _tableConfig.getTableName();
-    LOGGER.info("Starting {} instance assignment for table {}", instancePartitionsName, tableNameWithType);
 
-    boolean minimizeDataMovement = instanceAssignmentConfig.isMinimizeDataMovement();
+    // minimizeDataMovement might be set back to false within InstanceTagPoolSelector and InstancePartitionSelector
+    // if existingInstancePartitions is null.
+    boolean minimizeDataMovement =
+        minimizeDataMovementOverride == null ? instanceAssignmentConfig.isMinimizeDataMovement()
+            : minimizeDataMovementOverride;
+    LOGGER.info("Starting {} instance assignment for table {}, instanceAssignmentConfig.isMinimizeDataMovement()={}, "
+            + "minimizeDataMovementOverride={}, minimizeDataMovement={}", instancePartitionsName, tableNameWithType,
```

Comment:
nit: it's not enforced but most log msgs use this convention `... minimizeDataMovementOverride: {}` instead of `=`
and better rename `instanceAssignmentConfig.isMinimizeDataMovement()=` to something more readable